File: Algorithms/443. String Compression/go/443_string_compression.go
package main

import "fmt"

func compress(chars []byte) int {
	i := 0
	ansindex := 0

	l := len(chars)

	for i < l {
		j := i
		for j < l && chars[j] == chars[i] {
			j++
		}
		freq := j - i

		chars[ansindex] = chars[i]
		ansindex++
		if freq > 1 {
			temp := fmt.Sprintf("%d", freq)

			for _, c := range []byte(temp) {
				chars[ansindex] = c
				ansindex++
			}
		}
		i = j
	}
	return ansindex
}

func main() {
	test1 := []byte{'a', 'a', 'b', 'b', 'c', 'c', 'c'}
	test2 := []byte{'a'}
	test3 := []byte{'a', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b', 'b'}

	fmt.Printf("%c\n %v\n", test1, compress(test1))
	fmt.Printf("%c\n %v\n", test2, compress(test2))
	fmt.Printf("%c\n %v\n", test3, compress(test3))
}

File: Algorithms/443. String Compression/python/443_string_compression.py
from typing import List

class Solution:
    def compress(self, chars: List[str]) -> int: 
        i = index = 0
        
        n = len(chars)
        while i < n:
            j=i
            while j<n and chars[j] == chars[i]:
                j += 1
            
            freq = j-i
            
            chars[index] = chars[i]
            index +=1
            
            if freq > 1:
                freq = str(freq)
                for val in freq:
                    chars[index] = val
                    index +=1
            i = j
            
        print(chars[:index])
        return index
    
sol = Solution()

test1 = ["a","a","b","b","c","c","c"]

test2 = ["a"]

test3 = ["a","b","b","b","b","b","b","b","b","b","b","b","b"]

print(sol.compress(test1))

print(sol.compress(test2))

print(sol.compress(test3))

